PERF: Fast check on equivalent arrays in PyArray_EQUIVALENTLY_ITERABLE_OVERLAP_OK

eendebakpt opened the pull request:
The `PyArray_EQUIVALENTLY_ITERABLE_OVERLAP_OK` has to go through all the checks for identical arrays (a common case for inplace operations). This PR adds a fast check to see whether the two arrays to be compared are identical.

Benchmark
```
import numpy as np
import time

x=np.random.rand(10,20)   
niter=20_0000

t0=time.perf_counter()
for ii in range(niter):
    x=np.cos(x, out=x)
dt=time.perf_counter()-t0
print(dt)
```
Results in:
```
main: 0.51
PR: 0.39
```

Addresses one of the items in #21455

<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      https://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      https://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      https://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

eendebakpt committed:
fast check on equivalent arrays in PyArray_EQUIVALENTLY_ITERABLE_OVERLAP_OK

eendebakpt committed:
fix for case stride1==0

eendebakpt commented:
The `stride1 !=0` part of the condition is to catch the case
```
    import numpy as np
    x = np.array(2)
    np.add(x, [1], x)
```

seberg commented:
I am not wondering if it may also influence the the case of (I admit, I hope nobody does this!):
```
arr = np.array([5])
arr = np.broadcast_to(arr, 10)
arr.flags.writeable = True

np.add(arr, 0, out=arr)
```
But probably only interesting for a potential test...

I think I would remove the `assert` here again, it will lead to hard crashes anyway ;).  Also wondering if just repeating that "macro" call may not read easier, but thats just styling now.

eendebakpt committed:
remove assert statements

eendebakpt commented:
> I am not wondering if it may also influence the the case of (I admit, I hope nobody does this!):
> 
> ```
> arr = np.array([5])
> arr = np.broadcast_to(arr, 10)
> arr.flags.writeable = True
> 
> np.add(arr, 0, out=arr)
> ```
> 
> But probably only interesting for a potential test...

Output of your example is the same for main as for this PR: an array `[5 5 5 5 5 5 5 5 5 5]`.

seberg commented on numpy/core/src/common/lowlevel_strided_loops.h:
```diff
@@ -688,10 +688,19 @@ PyArray_EQUIVALENTLY_ITERABLE_OVERLAP_OK(PyArrayObject *arr1, PyArrayObject *arr
+
+    if (arr1 == arr2 && stride1 !=0) {
+        // case common for inplace operations
+        return 1;
```
`PyArray_TRIVIAL_PAIR_ITERATION_STRIDE` returns 0 for 1-sized arrays (i.e. 0-D).  So we miss those cases here. OTOH, 0-D arrays are (currently) pretty rare.

Maybe we should just add that logic and move it into the return?  When `(arr1 == arr2)`, we can `return size <= 1 || stride != 0;` (`solve_may_share_memory` really does nothing in the whole `arr1 == arr2` branch.)

Another thought, is to use `PyArray_DATA(arr1) == PyArray_DATA(arr2)`, we need the data anyway and it will allow catching views (may also help subclasses for example).

I agree with the change! I am just still a bit curious if it would not make more sense to push it into `solve_may_share_memory` so that it is used everywhere.

Another point is, that th `stride1 != 0` check here doesn't really seem right unless we care about my weird example, the check here really protects us from incorrect shape checking earlier... 

Anyway, if nothing moves soon, I may just add a brief comment and merge then.  But if you have an idea or think moving it is a good idea please do.  Otherwise, the only thing would be style nitpicks (blank line and space after !=).

eendebakpt replied:
I guess it depends on the use cases of interest (how often they occur and whether they are time critical). E.g. `PyArray_DATA(arr1) == PyArray_DATA(arr2)` will catch more cases, but perhaps it does require additional checks (e.g. on the stride or other properties?).

Maybe open an issue with label 'good first issue' to further investigate ideas like moving the check into `solve_may_share_memory ` or using `return size <= 1 || stride != 0`?

eendebakpt committed:
whitespace

seberg commented:
The code clutter is minimal now, so I am OK with this.  But to confirm that you think it is still worthwhile?

seberg commented:
Wait, I posted this on the wrong PR...  I meant to ask for the `int` change one.

seberg committed:
DOC: Add comment that output broadcasting is rejected by overlap detection

That is a bit of a weird dynamic here, and it would probably nice to clean
up.  However, it is also not a reason to not add the fast-path right now.

seberg commented:
@eendebakpt thanks!  Lets put this in.  I added (a longish) comment on the weird subtlety that this the stride check seems to implicitly reject output broadcasting.
